### PR TITLE
Robbe we miss you

### PR DESCRIPTION
**Pull request by timodw**
First controller implementation, not yet tested

### CODE REVIEW

File: src/controller/controller.ino [hunk 1]
```diff
@@ -27,19 +34,38 @@
...
+  uint8_t type;
+};
+
 MCP2515 mcp2515(10);
 
 uint8_t state = STATE_INACTIVE;
+struct module connected_modules_ids[OBUS_MAX_MODULES];
+uint8_t nr_connected_modules;
+uint8_t strikeouts;
+uint8_t game_running;
```
Comment (redfast00) [inline src/controller/controller.ino:48]: Make this a boolean? Or if there need to be more states in the future, an enum?

File: src/controller/controller.ino [hunk 2]
```diff
@@ -53,16 +79,204 @@ uint16_t make_id(uint8_t id, bool priority, uint8_t type) {
 		(uint16_t) id;
 }
 
+
+struct module get_module_info(uint16_t can_id) {
+  uint8_t module_type = can_id & 0x0300;
```
Comment (redfast00) [inline src/controller/controller.ino:84]: This does not fit in a `uint8_t`, did you forget to shitf?

File: src/controller/controller.ino [hunk 3]
```diff
@@ -53,16 +79,204 @@ uint16_t make_id(uint8_t id, bool priority, uint8_t type) {
...
+  uint8_t module_id = can_id & 0x00FF;
+
+  struct module module_info;
+  module_info.type = module_type;
+  module_info.id = module_id;
+  return module_info;
+}
+
+
+uint8_t check_solved() {
```
Comment (redfast00) [inline src/controller/controller.ino:94]: Return a boolean?

File: src/controller/controller.ino [hunk 4]
```diff
@@ -53,16 +79,204 @@ uint16_t make_id(uint8_t id, bool priority, uint8_t type) {
...
+  uint8_t solved = 1;
+  for (int i; i<32; i++) {
+    if (unsolved_puzzles != 0) {
```
Comment (redfast00) [inline src/controller/controller.ino:97]: `i` is not used inside the loop?

File: src/controller/controller.ino [hunk 5]
```diff
@@ -53,16 +79,204 @@ uint16_t make_id(uint8_t id, bool priority, uint8_t type) {
...
+    }
+  } else if (current_time - hello_round_start > OBUS_DISC_DURATION * 1000) {
+    state = STATE_GAME;
+    Serial.println("End of discovery round");
+  }
 }
 
+
+void initialize_game() {
+  strikeouts = 0;
```
Comment (redfast00) [inline src/controller/controller.ino:186]: Suggest naming this `strikes` (a strike is what you get when you make an error, a strikeout is what you get when you have too many strikes and the bomb explodes (cfr. timeout)

File: src/controller/controller.ino [hunk 6]
```diff
@@ -53,16 +79,204 @@ uint16_t make_id(uint8_t id, bool priority, uint8_t type) {
...
+
+  uint16_t game_duration_millis = (uint16_t) OBUS_GAME_DURATION * 1000;
+
+  uint8_t message[OBUS_MSG_LENGTH];
+  message[0] = OBUS_MSGTYPE_C_GAMESTART;
+  message[1] = (uint8_t) ((game_duration_millis & 0xFF000000) >> 0x18);
+  message[2] = (uint8_t) ((game_duration_millis & 0x00FF0000) >> 0x10);
+  message[3] = (uint8_t) ((game_duration_millis & 0x0000FF00) >> 0x08);
+  message[4] = (uint8_t) (game_duration_millis & 0x000000FF);
```
Comment (redfast00) [inline src/controller/controller.ino:195]: Nice! I like that you didn't assume endianess and directly copied the `uint32_t` into the buffer.

**Comment by Robbe7730** [comment]
❤️

**Pull request merged**